File: leetcode/python/11.container-with-most-water.py
from typing import List

class Solution:
    def maxArea(self, height: List[int]) -> int:
        res = 0
        l = 0
        r = len(height) - 1
        while l < r:
            t = min(height[l], height[r]) * (r - l)
            if t > res:
                res = t
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1

        return res

File: leetcode/cpp/11.container-with-most-water.cpp
#include "oj_header.h"

class Solution
{
public:
    int maxArea(vector<int> &height)
    {
        int l = 0, r = height.size() - 1;
        // 双指针 移动短的指针有可能会更大 而移动更长的指针肯定会更小
        int res = 0;
        while (l < r) {
            int t = (r - l) * min(height[l], height[r]);
            res = max(res, t);
            if (height[l] < height[r]) {
                ++l;
            } else {
                --r;
            }
        }
        return res;
    }
};
